filters/builtin/mod_header.go

```diff
@@ -0,0 +1,80 @@
+package builtin
+
+import (
+	"regexp"
+
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/filters"
+)
+
+type modRequestHeader struct {
+	headerName  string
+	rx          *regexp.Regexp
+	replacement string
+	template    *eskip.Template
+}
+
+// Returns a new modRequestHeader filter Spec, whose instances execute
+// regexp.ReplaceAllString on the request host. Instances expect two
+// parameters: the expression to match and the replacement string.
+// Name: "modRequestHeader".
+func NewModRequestHeader() filters.Spec { return &modRequestHeader{} }
+
+func (spec *modRequestHeader) Name() string {
+	return ModRequestHeaderName
+}
+
+// Creates instances of the modRequestHeader filter.
+//
+//lint:ignore ST1016 "spec" makes sense here and we reuse the type for the filter
+func (spec *modRequestHeader) CreateFilter(config []interface{}) (filters.Filter, error) {
+	if len(config) != 3 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	headerName, ok := config[0].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	expr, ok := config[1].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	replacement, ok := config[2].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	rx, err := regexp.Compile(expr)
+	if err != nil {
+		return nil, err
+	}
+
+	return &modRequestHeader{headerName: headerName, rx: rx, replacement: replacement}, nil
+}
+
+// Modifies the host with regexp.ReplaceAllString.
+func (f *modRequestHeader) Request(ctx filters.FilterContext) {
+	req := ctx.Request()
+
+	if f.headerName == "Host" {
+		h := req.Header.Get("Host")
+
+		if h == "" && req.Host != "" {
+			req.Host = f.rx.ReplaceAllString(req.Host, f.replacement)
```

Comment:
we have the SetOutgoingHost() on the filter context. Do you think it would work here, too? See: https://github.com/zalando/skipper/blob/master/filters/builtin/headerfilter.go#L153

and the SetOutgoingHost() documentation here: https://godoc.org/github.com/zalando/skipper/filters#FilterContext

I think using the SetOutgoingHost() will also make it consistent with the behavior of the -proxy-preserve-host global option and the preserveHost() predicate.

Reply:
Applied your feedback. WDYT?

Reply:
lgtm!